akka-stream-tests/src/test/scala/akka/stream/impl/fusing/InterpreterSupervisionSpec.scala

```diff
@@ -0,0 +1,656 @@
+      lastEvents() should be(Set(RequestOne, OnNext(1)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(-1)))
+
+      upstream.onNext(2) // boom
+      lastEvents() should be(Set(RequestOne))
+
+      downstream.requestOne()
+      lastEvents() should be(Set.empty)
+
+      upstream.onNext(3)
+      lastEvents() should be(Set(RequestOne, OnNext(3)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(-3)))
+    }
+
+    "resume when Expand `seed` throws" in new TestSetup(Seq(Expand(
+      (in: Int) ⇒ if (in == 2) throw TE else in,
+      (agg: Int) ⇒ (agg, -math.abs(agg)),
+      resumingDecider))) {
+
+      lastEvents() should be(Set(RequestOne))
+
+      upstream.onNext(1)
+      lastEvents() should be(Set.empty)
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne, OnNext(1)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(-1)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(-1)))
+
+      upstream.onNext(2) // boom
+      lastEvents() should be(Set(RequestOne))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(-1)))
+
+      upstream.onNext(3)
+      lastEvents() should be(Set.empty)
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne, OnNext(3)))
+    }
+
+    "restart when Expand `extrapolate` throws" in new TestSetup(Seq(Expand(
+      (in: Int) ⇒ in,
+      (agg: Int) ⇒ if (agg == 2) throw TE else (agg, -math.abs(agg)),
+      restartingDecider))) {
+
+      lastEvents() should be(Set(RequestOne))
+
+      upstream.onNext(1)
+      lastEvents() should be(Set.empty)
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne, OnNext(1)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(-1)))
+
+      upstream.onNext(2) // boom
+      lastEvents() should be(Set.empty)
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne))
+
+      upstream.onNext(3)
+      lastEvents() should be(Set.empty)
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne, OnNext(3)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(-3)))
+    }
+
+    "resume when Expand `extrapolate` throws" in new TestSetup(Seq(Expand(
+      (in: Int) ⇒ in,
+      (agg: Int) ⇒ if (agg == 2) throw TE else (agg, -math.abs(agg)),
+      resumingDecider))) {
+
+      pending // see blow FIXME
+
+      lastEvents() should be(Set(RequestOne))
+
+      upstream.onNext(1)
+      lastEvents() should be(Set.empty)
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne, OnNext(1)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(-1)))
+
+      upstream.onNext(2) // boom
+      lastEvents() should be(Set.empty)
+
+      downstream.requestOne() // FIXME this throws IllegalStateException: Cannot pull while holding, only pushAndPull
```

Comment:
yeah, if it is in holding state it means that the one upstream and one downstream token is "owned" by the stage. In this case it *must* call a pushAndPull. There is no way around this. (Please note that this is not interpreter specific, this will be the same issue for a hand-crafted Processor. This is true for the other FIXMEs as well)

The same problem will repeat many times: in order to keep boundedness and not lose request tokens, a pull *must* be answered by a push, which unfortunately means to provide an element upon failure.